[extension/jaegerremotesampling] Support reload_interval option in remote

zcross opened the pull request:
**Description:** <Describe what has changed.>
Updates the jaegerremotesampling extension's `remote` mode of usage to support the `reload_interval` caching option already supported in `file` mode of usage.

**Link to tracking Issue:** #24840

**Testing:** 
* Integration test updates show collector<>remote interaction with and without caching
* Unit tests cover specific caching behavior inclusive of concurrency (leak detector / race detector etc)

**Documentation:** Existing readme snippet referring to the `reload_interval` config option explains that it works for both modes of `remote`.

zcross committed:
[extension/jaegerremotesampling] Support reload_interval option in remote mode with naive TTL cache

zcross commented:
cc @atoulme who has some context via the prior PR #24414

zcross commented on extension/jaegerremotesampling/README.md:
```diff
@@ -24,7 +24,7 @@ Note that the port `14250` will clash with the Jaeger Receiver. When both are us
 
 Although this extension is derived from Jaeger, it can be used by any clients who can consume this standard, such as the [OpenTelemetry Java SDK](https://github.com/open-telemetry/opentelemetry-java/tree/v1.9.1/sdk-extensions/jaeger-remote-sampler).
 
```
This README diff captures the essence of the PR (and reflects my intent to use the same config field/language rather than going into the nuanced differences)

zcross commented on extension/jaegerremotesampling/internal/remote_strategy_cache.go:
```diff
@@ -0,0 +1,123 @@
+	}
+}
+
+// periodicallyClearCache provides a secondary "cleanup" mechanism motivated by wanting to guarantee
```
I'm open to feedback about this. I don't love it in this scheduled form. We could just "hint" it from client requests and launch a go routine that does this same work depending on some synchronized state (last clear time, etc)

frzifus commented:
Thats a great enhancement - But does it make sense to resolve the [jaeger dependency issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18401) first?

Upgrading it, seems to be easy. Right now, I try to hide things behind a feature flag ([PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18485)).

zcross commented:
> Thats a great enhancement - But does it make sense to resolve the [jaeger dependency issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18401) first?
> 
> Upgrading it, seems to be easy. Right now, I try to hide things behind a feature flag ([PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18485)).

I'll take a look at your PR – I'm still catching up on the "why" behind it and trying to understand if it's fixing some issue currently impacting clients or just a nice "let's unbreak a dep upgrade" change.

jpkrohling commented on extension/jaegerremotesampling/internal/remote_strategy_cache_test.go:
```diff
@@ -0,0 +1,247 @@
+	// whitebox assert for internal timestamp tracking (we don't want a caching bug manifesting as stale data serving)
+	// (post-write)
+	assert.Equal(t, serviceStrategyCacheEntry{
+		retrievedAt:      testTime,
```
```suggestion
		retrievedAt:      firstWriteTime,
```

So that this assertion will be the same as the one performed under "reading does not mutate internal cache state"?

zcross replied:
good call for clarity for readers

jpkrohling commented on extension/jaegerremotesampling/internal/remote_strategy_cache_test.go:
```diff
@@ -0,0 +1,247 @@
+	assert.False(t, ok)
+	assert.Nil(t, result)
+
+	assert.Equal(t, serviceStrategyCacheEntry{
```
Interesting: this means that the item is still in the cache, it's just not returned when `get` is called.

zcross replied:
Yeah, I could also just delete stale items on stale reads. I prefer to just let cache misses replace them.
This is sort of why I created a "cleaner" in a periodic goroutine to ensure that the memory footprint doesn't grow unbounded, rather than 100% relying on queries managing to clean up stale items (because queries are for a single service).

jpkrohling commented on extension/jaegerremotesampling/internal/remote_strategy_cache_test.go:
```diff
@@ -0,0 +1,247 @@
+	assert.Nil(t, result)
+	result, ok = cache.get("barSvc")
+	assert.True(t, ok)
+	assert.Equal(t, testStrategyResponseB, result)
```
I'm confused: why is this B, and not A?

zcross replied:
good catch, this is actually the result of my test fixtures being identical (as you pointed out in an earlier comment). fixing this now

jpkrohling commented on extension/jaegerremotesampling/internal/remote_strategy_store.go:
```diff
@@ -27,15 +30,30 @@ type grpcRemoteStrategyStore struct {
+	}
+	freshResult, err := g.delegate.GetSamplingStrategy(g.enhanceContext(ctx), serviceName)
+	if err != nil {
+		return nil, fmt.Errorf("GetSamplingStrategy call to remote failed: %s", err.Error())
```
```suggestion
		return nil, fmt.Errorf("GetSamplingStrategy call to remote failed: %w", err)
```

zcross commented on extension/jaegerremotesampling/internal/remote_strategy_cache.go:
```diff
@@ -0,0 +1,123 @@
+	return nil
+}
+
+type noopStrategyCache struct{}
```
This is used when the cache is not configured for use (`reload_interval` = 0 --> no need to interact with an actual cache at all with a zero-valued TTL).

zcross commented on extension/jaegerremotesampling/internal/remote_strategy_cache_test.go:
```diff
@@ -0,0 +1,247 @@
+
+var testStrategyResponseB = &sampling.SamplingStrategyResponse{
+	ProbabilisticSampling: &sampling.ProbabilisticSamplingStrategy{
+		SamplingRate: 0.1337,
```
whoops, that was definitely my intent (to have different contents, not just identities). thanks!

zcross commented on extension/jaegerremotesampling/internal/remote_strategy_cache.go:
```diff
@@ -0,0 +1,123 @@
+type serviceStrategyCache interface {
+	get(serviceName string) (*sampling.SamplingStrategyResponse, bool)
+	put(serviceName string, response *sampling.SamplingStrategyResponse)
+	close() error
```
whoops, addressing this now. after rebasing I managed to actually throw away the invocation of this closer, so I'm glad this came up.

BTW: since we use the Jaeger-defined `strategystore.StrategyStore` interface for both our remote/local variations, I'm going to just be returning this cache now that it's an `io.Closer` for easy access (to close during component lifecycle) in `extension.go`

zcross commented:
Hey, I'm sorry to do this but I'm going to close and open a new PR to hopefully re-trigger EasyCLA checks. I created my LFX profile recently and want to make sure that it's detecting myself / my org / etc correctly. I have, however, addressed the feedback (so far) on this PR. Thanks all!

zcross commented:
(Continued in #24981 after some EasyCLA double-checking which is good as expected; note that I'm not dismissing the fact that we want to deal with the thrift->proto change that Bene raised)
